src/state/slices/opportunitiesSlice/opportunitiesSlice.ts

```diff
@@ -81,70 +86,115 @@ export const opportunitiesApi = createApi({
   endpoints: build => ({
     getOpportunityMetadata: build.query<GetOpportunityMetadataOutput, GetOpportunityMetadataInput>({
       queryFn: async ({ opportunityId, opportunityType, defiType }, { dispatch, getState }) => {
-        const resolver = getMetadataResolversByDefiProviderAndDefiType(
-          DefiProvider.FoxFarming,
-          defiType,
+        const maybeResolver = O.toNullable(
+          getMetadataResolversByDefiProviderAndDefiType(DefiProvider.FoxFarming, defiType),
         )
-        const resolved = await resolver({
-          opportunityId,
-          opportunityType,
-          reduxApi: { dispatch, getState },
-        })
 
-        dispatch(opportunities.actions.upsertOpportunityMetadata(resolved.data))
+        if (!maybeResolver) {
+          throw new Error(`resolver for ${DefiProvider.FoxFarming}::${defiType} not implemented`)
+        }
+
+        const maybeData = await pipe(
+          maybeResolver,
+          resolver =>
+            TE.tryCatch(
+              () =>
+                resolver({
+                  opportunityId,
+                  opportunityType,
+                  reduxApi: { dispatch, getState },
+                }),
+              E.toError,
+            ),
+          TE.map(resolved => resolved.data),
+        )()
+
+        return pipe(
+          maybeData,
+          E.match<
+            Error,
+            GetOpportunityMetadataOutput,
+            { data: GetOpportunityMetadataOutput } | { error: BaseQueryError<any> }
+          >(
+            err => ({
+              error: {
+                error: err.message,
+                status: 'CUSTOM_ERROR',
+              },
+            }),
+            data => {
+              dispatch(opportunities.actions.upsertOpportunityMetadata(data))
 
-        return { data: resolved.data }
+              return { data }
+            },
+          ),
+        )
```

Comment:
There is no need to unwrap our `maybeResolver`, we can (and should) leave it as an `Option`.
Then we can just handle the `None` case using a `match` in our `pipe`:

```ts
const maybeResolver = getMetadataResolversByDefiProviderAndDefiType(
  DefiProvider.FoxFarming,
  defiType,
)

const maybeData = await pipe(
  maybeResolver,
  O.match(
    () => {
      throw new Error(
        `resolver for ${DefiProvider.FoxFarming}::${defiType} not implemented`,
      )
    },
    resolver =>
      TE.tryCatch(
        () =>
          resolver({
            opportunityId,
            opportunityType,
            reduxApi: { dispatch, getState },
          }),
        E.toError,
      ),
  ),
  TE.map(resolved => resolved.data),
)()

return pipe(
  maybeData,
  E.match<
    Error,
    GetOpportunityMetadataOutput,
    { data: GetOpportunityMetadataOutput } | { error: BaseQueryError<any> }
    >(
    err => ({
      error: {
        error: err.message,
        status: 'CUSTOM_ERROR',
      },
    }),
    data => {
      dispatch(opportunities.actions.upsertOpportunityMetadata(data))

      return { data }
    },
  ),
)
```

Reply:
Some more thoughts here:

The `thow` in my suggestion here is poor practice, as we should never throw in a functional program.

Instead, we should return an error, i.e. an `Either` monad with `<Error, HappyPathType>`.

One option here is to do something like:

```ts
const maybeData = await pipe(
  maybeResolver,
  O.chainNullableK(resolver =>
    TE.tryCatch(
      () =>
        resolver({
          opportunityId,
          opportunityType,
          reduxApi: { dispatch, getState },
        }),
      E.toError,
    ),
  ),
)
```

Then we need to unwrap it from the `Option`, maybe by `getOrElse`ing it, where the "or else" is an `Either<Error, Never>`.

I'm not yet sure exactly how to achieve that in `fp-ts`, but that's the conceptual approach.

Reply:
Mixing `TaskEither` and `Option` wasn't an easy one, but looks like this is the guy 🤞🏽
21ad189129ae3d19ca40b4b4a2ec9086172c1b72
Managed to carry the optionals all the way through the pipeline, make sure the right errors are "caught" (or rather carried) in the right places as `Left`s, gave it some testing locally with breakpoints and the flow is still the same!